web/bridge_types_controller_test.go

```diff
@@ -40,7 +40,7 @@ func TestBridgeTypesController_Create_AdapterExistsError(t *testing.T) {
 		"application/json",
 		bytes.NewBuffer(cltest.LoadJSON("../internal/fixtures/web/existing_core_adapter.json")),
 	)
-	cltest.CheckStatusCode(t, resp, 400)
+	cltest.CheckStatusCode(t, resp, 500)
```

Comment:
This should remain `400`. Creating an adapter that already exists and receiving an error about it isn't an internal error 500. I did leave the `500` in my sample code snippet, my fault.

Reply:
Returning 400 makes sense when the user tries to add a duplicate bridge type. Returning 500 makes sense when adding the bridge type to the database fails. There is currently no way to handle both cases without passing in the gin.Context to the validation method. Are you ok with having both error cases return 400?

Reply:
Great point. We want to discern between the two, so now could be a good time to introduce a `ValidationError` type in `validators.go` that complies with [the `error` interface](https://blog.golang.org/error-handling-and-go). Then in the controller have

```
func (btc *BridgeTypesController) Create(c *gin.Context) {
	bt := &models.BridgeType{}

	if err := c.ShouldBindJSON(bt); err != nil {
		c.JSON(500, gin.H{
			"errors": []string{err.Error()},
		})
	} else if err = services.ValidateAdapter(bt, btc.App.GetStore()); err != nil {
		c.JSON(StatusCodeForError(err), gin.H{ // new method in router.go?
			"errors": []string{err.Error()},
		})
```

This StatusCodeForError could do type assertion.

We'll add this in a subsequent commit, unless you're up for tackling it now. Which would you prefer?

Reply:
I'll tackle it

Reply:
Great, thanks!